src/js/components/FocusedContainer.js

```diff
@@ -12,13 +30,112 @@ export const FocusedContainer = ({
 }) => {
   const [bodyOverflowStyle, setBodyOverflowStyle] = useState('');
   const ref = useRef(null);
+  const preNodeRef = useRef(null);
+  const postNodeRef = useRef(null);
+
+  const { contextValue, hasRoots } = useRoots();
+  const { roots: contextRoots } = contextValue;
+  // When attemptFocus moves focus around to find the focusable element,
+  // set this to true so handleTrapFocus doesn't focus the element.
+  const ignoreUtilFocusRef = useRef(false);
 
-  const roots = useContext(RootsContext);
-  const [nextRoots, setNextRoots] = useState(roots);
   useEffect(() => {
-    // make sure value of null is not added to array
-    if (ref.current) setNextRoots([...roots, ref.current]);
-  }, [roots]);
+    const container = ref.current;
+    const roots = contextRoots.current;
+
+    const attemptFocus = (element) => {
+      // Check if the element is focusable; if not, return false
+      if (!isFocusable(element)) {
+        return false;
+      }
+      ignoreUtilFocusRef.current = true;
+      try {
+        element.focus();
+      } catch (e) {
+        // continue regardless of error
+      }
+      ignoreUtilFocusRef.current = false;
+      // Return true if the element is currently the active element or has focus
+      return document.activeElement === element;
+    };
+
+    const focusFirstDescendant = (element) => {
+      // Iterate through all child nodes of the provided element
+      for (let i = 0; i < element.childNodes.length; i += 1) {
+        const child = element.childNodes[i];
+        if (attemptFocus(child) || focusFirstDescendant(child)) return true;
+      }
+      // If no focusable child or descendant was found, return false
+      return false;
+    };
+
+    const focusLastDescendant = (element) => {
+      for (let i = element.childNodes.length - 1; i >= 0; i -= 1) {
+        const child = element.childNodes[i];
+        if (attemptFocus(child) || focusLastDescendant(child)) return true;
+      }
+      return false;
+    };
+
+    const handleTrapFocus = (e) => {
+      if (
+        !hidden &&
+        trapFocus &&
+        ignoreUtilFocusRef.current === false &&
+        container &&
+        // only perform focus if this is the most recently opened drop
+        roots[roots.length - 1] === container
+      ) {
+        if (container.contains(e.target)) {
+          container.lastFocus = e.target;
+        } else {
+          focusFirstDescendant(container);
+          if (container.lastFocus === document.activeElement) {
+            focusLastDescendant(container);
+          }
+          container.lastFocus = document.activeElement;
+        }
+      }
+    };
+
+    const addListeners = () => {
+      document.addEventListener('focus', handleTrapFocus, true);
+    };
+
+    const removeListeners = () => {
+      document.removeEventListener('focus', handleTrapFocus, true);
+    };
```

Comment:
Since these are trivial and only used on one spot you could just inline them where used and avoid creating functions as well as simplify this already large useEffect

Reply:
Agree, will update.